### PR TITLE
8265795: vmTestbase/nsk/jvmti/AttachOnDemand/attach022/TestDescription.java fails when running with JEP 416

### PR DESCRIPTION
**Pull request by lmesnik**
The VMObjectAlloc jvmti event was not generated for objects created using MethodHanldle. The fix adds posting of the event into Unsafe_AllocateInstance.

While fixing this bug I noticed that event is not posted in the intrinsics version for many functions where it is used. Including  but not limited to clone(), invoke()m allocateInstance() and allocateUninitializedArray(). There are might be other intensified functions (not analogs JVM_ENTRY versions) that allocate objects without post events. 

I think it is needed to implement some common way to handle this and cover it in another issue.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265795](https://bugs.openjdk.java.net/browse/JDK-8265795): vmTestbase/nsk/jvmti/AttachOnDemand/attach022/TestDescription.java fails when running with JEP 416

### Reviewers
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to e160dbe3eafe82c83319d0bb1baab5702686aef9
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to e160dbe3eafe82c83319d0bb1baab5702686aef9

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6478/head:pull/6478` \
`$ git checkout pull/6478`

Update a local copy of the PR: \
`$ git checkout pull/6478` \
`$ git pull https://git.openjdk.java.net/jdk pull/6478/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6478`

View PR using the GUI difftool: \
`$ git pr show -t 6478`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6478.diff">https://git.openjdk.java.net/jdk/pull/6478.diff</a>

</details>

### CODE REVIEW

**Review by sspitsyn — APPROVED** [review]
Hi Leonid,
This fix looks good to me.
Thanks,
Serguei

File: test/hotspot/jtreg/serviceability/jvmti/VMObjectAlloc/VMObjectAllocTest.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    public static void main(String[] args) throws Throwable {
+
+        MethodHandles.Lookup publicLookup = MethodHandles.publicLookup();
+        MethodType mt = MethodType.methodType(void.class, String.class);
+        MethodHandle mh = publicLookup.findConstructor(VMObjectAllocTest.class, mt);
+        mh.invoke("str");
+
+        if(getNumberOfAllocation() != 1) {
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/serviceability/jvmti/VMObjectAlloc/VMObjectAllocTest.java:49]: space after 'if' please

File: test/hotspot/jtreg/serviceability/jvmti/VMObjectAlloc/libVMObjectAlloc.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  err = jvmti->SetEventNotificationMode(JVMTI_ENABLE, JVMTI_EVENT_VM_OBJECT_ALLOC , NULL);
+  if (err != JVMTI_ERROR_NONE) {
+    return JNI_ERR;
+  }
+
+  return JNI_OK;
+}
+
+}
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/serviceability/jvmti/VMObjectAlloc/libVMObjectAlloc.cpp:91]: This looks spurious ??

**Review by dholmes-ora — APPROVED** [review]
Hi Leonid,

Functional fix looks good. A couple of minor nits below.

I agree that fixing intrinsics should be a separate issue - I have to worry that the overhead of posting events can dwarf the operation itself. I would guess the intrinsic would need a short-cut to check if the event is enabled and if so drop back to non-intrinsic version.

Thanks,
David

File: test/hotspot/jtreg/serviceability/jvmti/VMObjectAlloc/VMObjectAllocTest.java (hunk 1 above)
Comment (lmesnik) [reply test/hotspot/jtreg/serviceability/jvmti/VMObjectAlloc/VMObjectAllocTest.java]: fixed

**Pull request closed without merging**